File: src/core/memory/session.ts
/**
 * 会话记忆管理
 */

import { db } from '@/lib/db';

interface Message {
    role: 'user' | 'assistant' | 'system';
    content: string;
    timestamp: number;
}

const sessionCache = new Map<string, Message[]>();
const MAX_MESSAGES = 20;

export async function addMessage(
    sessionId: string,
    role: 'user' | 'assistant' | 'system',
    content: string
): Promise<void> {
    if (!sessionCache.has(sessionId)) {
        sessionCache.set(sessionId, []);
    }
    const messages = sessionCache.get(sessionId)!;
    messages.push({ role, content, timestamp: Date.now() });

    if (messages.length > MAX_MESSAGES) {
        sessionCache.set(sessionId, messages.slice(-MAX_MESSAGES));
    }

    try {
        const existingSession = await db.session.findUnique({ where: { id: sessionId } });
        if (!existingSession) {
            await db.session.create({
                data: { id: sessionId, domain: 'medical' },
            });
        }

        await db.message.create({
            data: {
                sessionId,
                role,
                content,
            },
        });
    } catch (e) {
        console.error('[SessionManager] 写入数据库失败:', e);
    }
}

export async function getHistory(sessionId: string): Promise<Message[]> {
    if (sessionCache.has(sessionId)) {
        return sessionCache.get(sessionId)!;
    }

    try {
        const messages = await db.message.findMany({
            where: { sessionId },
            orderBy: { createdAt: 'asc' },
            take: MAX_MESSAGES,
        });

        const result = messages.map(m => ({
            role: m.role as 'user' | 'assistant' | 'system',
            content: m.content,
            timestamp: new Date(m.createdAt).getTime(),
        }));

        sessionCache.set(sessionId, result);
        return result;
    } catch (e) {
        console.error('[SessionManager] 读取数据库失败:', e);
        return [];
    }
}

export async function buildHistoryContext(sessionId: string): Promise<string> {
    const history = await getHistory(sessionId);
    if (history.length === 0) return '';

    let context = '【对话历史】\n';
    for (const msg of history) {
        const roleText = msg.role === 'user' ? '用户' : '助手';
        context += `${roleText}：${msg.content}\n`;
    }

    return context;
}

export async function updateLastAssistant(sessionId: string, content: string): Promise<void> {
    const messages = sessionCache.get(sessionId);
    if (messages) {
        for (let i = messages.length - 1; i >= 0; i--) {
            if (messages[i].role === 'assistant') {
                messages[i].content = content;
                break;
            }
        }
    }

    try {
        const lastMsg = await db.message.findFirst({
            where: { sessionId, role: 'assistant' },
            orderBy: { createdAt: 'desc' },
        });
        if (lastMsg) {
            await db.message.update({
                where: { id: lastMsg.id },
                data: { content },
            });
        }
    } catch (e) {
        console.error('[SessionManager] 更新数据库失败:', e);
    }
}

export async function clearSession(sessionId: string): Promise<void> {
    sessionCache.delete(sessionId);
    try {
        await db.message.deleteMany({ where: { sessionId } });
        await db.session.delete({ where: { id: sessionId } });
    } catch (e) {
        console.error('[SessionManager] 清除会话失败:', e);
    }
}

export async function createSession(domain: string = 'medical', title?: string): Promise<string> {
    const session = await db.session.create({
        data: { domain, title: title || null },
    });
    sessionCache.set(session.id, []);
    return session.id;
}

export async function updateSessionTitle(sessionId: string, title: string): Promise<void> {
    try {
        await db.session.update({
            where: { id: sessionId },
            data: { title },
        });
    } catch (e) {
        console.error('[SessionManager] 更新会话标题失败:', e);
    }
}

export async function getUserSessions(userId?: string) {
    try {
        return await db.session.findMany({
            where: userId ? { userId } : undefined,
            orderBy: { updatedAt: 'desc' },
            take: 50,
        });
    } catch (e) {
        console.error('[SessionManager] 获取会话列表失败:', e);
        return [];
    }
}

File: server/core/memory/session.py
"""
Session memory management.
"""
import uuid
from datetime import datetime
from typing import Dict, Any, List, Optional

_sessions: Dict[str, Dict[str, Any]] = {}


async def create_session(domain: str = "medical", title: Optional[str] = None) -> str:
    session_id = str(uuid.uuid4())[:8]
    _sessions[session_id] = {
        "id": session_id,
        "domain": domain,
        "title": title or "新对话",
        "messages": [],
        "created_at": datetime.now().isoformat(),
        "updated_at": datetime.now().isoformat(),
    }
    return session_id


async def add_message(session_id: str, role: str, content: str) -> bool:
    session = _sessions.get(session_id)
    if not session:
        return False
    session["messages"].append({"role": role, "content": content, "timestamp": datetime.now().isoformat()})
    session["updated_at"] = datetime.now().isoformat()
    return True


async def get_history(session_id: str, limit: int = 50) -> List[Dict[str, Any]]:
    session = _sessions.get(session_id)
    if not session:
        return []
    return session["messages"][-limit:]


async def update_last_assistant(session_id: str, content: str) -> bool:
    session = _sessions.get(session_id)
    if not session:
        return False
    for msg in reversed(session["messages"]):
        if msg["role"] == "assistant":
            msg["content"] = content
            session["updated_at"] = datetime.now().isoformat()
            return True
    return False


async def update_session_title(session_id: str, title: str) -> bool:
    session = _sessions.get(session_id)
    if not session:
        return False
    session["title"] = title
    session["updated_at"] = datetime.now().isoformat()
    return True


async def clear_session(session_id: str) -> bool:
    if session_id in _sessions:
        del _sessions[session_id]
        return True
    return False


async def get_session(session_id: str) -> Optional[Dict[str, Any]]:
    return _sessions.get(session_id)


async def list_sessions(domain: Optional[str] = None) -> List[Dict[str, Any]]:
    sessions = list(_sessions.values())
    if domain:
        sessions = [s for s in sessions if s["domain"] == domain]
    return sorted(sessions, key=lambda x: x["updated_at"], reverse=True)